File: c++/balancedBinaryTree#110/try1.cpp
// This is the optimal solution

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) 
    {
        if(!root)
        {
            return true;
        }
        isBalance = true;

        compareHeight(root);
        return isBalance;
    }

private:

    int compareHeight(TreeNode* root)
    {
        if(!root)
        {
            return 0;
        }
        
        int leftTree = compareHeight(root->left);
        int rightTree = compareHeight(root->right);

        if(abs(leftTree - rightTree) > 1)
        {
            isBalance = false;
        }

        return 1 + max(leftTree, rightTree);
    }

    bool isBalance;
};

// Time: O(n)
// Space: O(n)

File: python/deleteNodesAndReturnForest#1110/try1.py
# This is the optimal solution

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def delNodes(self, root: Optional[TreeNode], to_delete: List[int]) -> List[TreeNode]:
        deleteSet = set(to_delete)
        roots = []
        def dfs(node, isRoot, isDeleted):
            if not node:
                return

            if node.val in deleteSet:
                isDeleted = True
                isRoot = False

            if isRoot:
                roots.append(node)
            
            leftChild, rightChild = node.left, node.right
            if leftChild and leftChild.val in deleteSet:
                node.left = None
            if rightChild and rightChild.val in deleteSet:
                node.right = None
            
            if isDeleted:
                dfs(leftChild, True, False)
                dfs(rightChild, True, False)
                del node
            else:
                dfs(leftChild, False, False)
                dfs(rightChild, False, False)   

        dfs(root, True, False)

        return roots

# Time: O(N)
# Space: O(h) / O(N)
